### ISSUE TITLE
如何在主干网络中加入自定义损失函数？

### ISSUE THREAD

**Issue by linshuchar** [open-mmlab/mmsegmentation#2128]
我需要在主干网络中加入两个模块，训练时要加入损失函数，看了代码好像损失都是在解码头那里

**Comment by xiexinch** [comment]
Hi @linshuchar,
MMSegmentation 实现的 EncoderDecoder 只能在解码头中加损失函数。如果您想在主干网络中加入自定义损失函数，可以自定义一个 CustomEncoderDecoder 以及 Backbone，重写 `train_step()` 方法的逻辑。

**Comment by linshuchar** [comment]
> Hi @linshuchar,
> 我定位到outputs = self.model.train_step(data_batch, self.optimizer, **kwargs)，训练时候返回的所有损失，是否我重新定义损失后返回给outputs就可以了？loss.backward()和优化器调整学习率的后续代码需要修改吗？具体实现在哪里呢？
> MMSegmentation 实现的 EncoderDecoder 只能在解码头中加损失函数。如果您想在主干网络中加入自定义损失函数，可以自定义一个 CustomEncoderDecoder 以及 Backbone，重写 `train_step()` 方法的逻辑。

**Comment by linshuchar** [comment]
我定位到outputs = self.model.train_step(data_batch, self.optimizer, **kwargs)，训练时候返回的所有损失，是否我重新定义损失后返回给outputs就可以了？loss.backward()和优化器调整学习率的后续代码需要修改吗？具体实现在哪里呢？

**Comment by xiexinch** [comment]
参数更新的实现在底层库 MMCV 里。

https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/iter_based_runner.py#L70

https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/optimizer.py#L61

**Comment by klyx9** [comment]
> 我定位到outputs = self.model.train_step（data_batch， self.optimizer， **kwargs），训练时候返回的所有损失，是否我重新定义损失后返回给outputs就可以了？loss.backward（）和优化器调整学习率的后续代码需要修改吗？具体实现在哪里呢？

请问你是这样做的吗？还是去修改mmcv底层？
